lib/MonadicRewrite.thy

```diff
@@ -14,24 +15,100 @@ imports
 
+lemma monadic_rewrite_inst:
+  "monadic_rewrite F E P f g \<Longrightarrow> monadic_rewrite F E P f g"
+  by assumption
+
+lemma monadic_rewrite_refl:
+  "monadic_rewrite F E \<top> f f"
+  by (simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_impossible:
+  "monadic_rewrite F E \<bottom> f g"
+  by (clarsimp simp: monadic_rewrite_def)
+
+lemma monadic_rewrite_trans:
+  "\<lbrakk> monadic_rewrite F E P f g; monadic_rewrite F E Q g h \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (P and Q) f h"
+  by (auto simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_trans_dup:
+  "\<lbrakk> monadic_rewrite F E P f g; monadic_rewrite F E P g h \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E P f h"
+  by (auto simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_imp:
+  "\<lbrakk> monadic_rewrite F E Q f g; \<And>s. P s \<Longrightarrow> Q s \<rbrakk> \<Longrightarrow> monadic_rewrite F E P f g"
+  by (auto simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_from_simple:
+  "P \<longrightarrow> f = g \<Longrightarrow> monadic_rewrite F E (\<lambda>_. P) f g"
+  by (simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_gen_asm:
+  "\<lbrakk> P \<Longrightarrow> monadic_rewrite F E Q f g \<rbrakk> \<Longrightarrow> monadic_rewrite F E ((\<lambda>_. P) and Q) f g"
+  by (auto simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_name_pre:
+  "\<lbrakk> \<And>s. P s \<Longrightarrow> monadic_rewrite F E ((=) s) f g \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E P f g"
+  by (auto simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_cases:
+  "\<lbrakk> P \<Longrightarrow> monadic_rewrite F E Q a b; \<not> P \<Longrightarrow> monadic_rewrite F E R a b \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (P \<longrightarrow> Q s) \<and> (\<not> P \<longrightarrow> R s)) a b"
+  by (cases P, simp_all)
+
+lemma monadic_rewrite_is_refl:
+  "x = y \<Longrightarrow> monadic_rewrite F E \<top> x y"
+  by (simp add: monadic_rewrite_refl)
+
+lemma monadic_rewrite_refl3:
+  "\<lbrakk> \<And>s. P s \<Longrightarrow> f s = g s \<rbrakk> \<Longrightarrow> monadic_rewrite F E P f g"
+  by (simp add: monadic_rewrite_def)
+
+lemmas monadic_rewrite_refl2 = monadic_rewrite_refl3[where P=\<top>]
+
+lemma monadic_rewrite_exists:
+  "(\<And>v. monadic_rewrite F E (Q v) m m')
+   \<Longrightarrow> monadic_rewrite F E ((\<lambda>s. \<forall>v. P v s \<longrightarrow> Q v s) and (\<lambda>s. \<exists>v. P v s)) m m'"
+  by (auto simp: monadic_rewrite_def)
+
+lemma monadic_rewrite_exists_v:
+  "\<lbrakk> \<And>v. monadic_rewrite E F (Q v) f g \<rbrakk>
+   \<Longrightarrow> monadic_rewrite E F (\<lambda>x. (\<exists>v. P v x) & (\<forall>v. P v x \<longrightarrow> Q v x)) f g"
+  by (rule monadic_rewrite_name_pre)
+     (fastforce elim: monadic_rewrite_imp)
+
+lemma monadic_rewrite_weaken:
+  "monadic_rewrite (F \<and> F') (E \<or> E') P f g
+   \<Longrightarrow> monadic_rewrite F' E' P f g"
```

Comment:
also for later: the name weaken is a bit strange for this rule. I get how it weakens, but for corres and others we called the `monadic_rewrite_imp` weaken. Probably better in the long run to be consistent there.

Reply:
What would you suggest? I'm not sure I follow exactly; these two weaken rules mess around with the failure flags.

Reply:
For this PR we shouldn't do anything yet, because it'd be a lot of renaming. Not even sure that it really is worth doing, it just bugs me a bit.

The point of inconsistency is that in corres and wp, the "weaken" rules refer to the preconditions + postcondition + return value relation. Here the "imp" rule does that, and the "weaken" rule refers to flags instead.

Reply:
No constructive suggestions, but I agree with Gerwin that the name `monadic_rewrite_weaken` leads to me expecting that the precondition is what will be weakened.